test/test_complex.py

```diff
@@ -1,45 +1,35 @@
-import math
 import torch
-from torch.testing._internal.common_utils import TestCase, run_tests, TEST_NUMPY
-import unittest
-
-if TEST_NUMPY:
-    import numpy as np
+from torch.testing._internal.common_device_type import instantiate_device_type_tests, dtypes
+from torch.testing._internal.common_utils import TestCase, run_tests
 
 devices = (torch.device('cpu'), torch.device('cuda:0'))
 
 
 class TestComplexTensor(TestCase):
-    def test_to_list_with_complex_64(self):
+    @dtypes(*torch.testing.get_all_complex_dtypes())
+    def test_to_list(self, device, dtype):
         # test that the complex float tensor has expected values and
         # there's no garbage value in the resultant list
-        self.assertEqual(torch.zeros((2, 2), dtype=torch.complex64).tolist(), [[0j, 0j], [0j, 0j]])
-
-    @unittest.skipIf(not TEST_NUMPY, "Numpy not found")
-    def test_exp(self):
-        def exp_fn(dtype):
-            a = torch.tensor(1j, dtype=dtype) * torch.arange(18) / 3 * math.pi
-            expected = np.exp(a.numpy())
-            actual = torch.exp(a)
-            self.assertEqual(actual, torch.from_numpy(expected))
+        self.assertEqual(torch.zeros((2, 2), device=device, dtype=dtype).tolist(), [[0j, 0j], [0j, 0j]])
 
-        exp_fn(torch.complex64)
-        exp_fn(torch.complex128)
-
-    def test_copy_real_imag_methods(self):
-        real = torch.randn(4)
-        imag = torch.randn(4)
+    @dtypes(torch.float32, torch.float64)
+    def test_copy_real_imag_methods(self, device, dtype):
+        real = torch.randn(4, device=device, dtype=dtype)
+        imag = torch.randn(4, device=device, dtype=dtype)
         complex_tensor = real + 1j * imag
         # TODO(#38095): Replace assertEqualIgnoreType. See issue #38095
         self.assertEqualIgnoreType(complex_tensor.copy_real(), real)
         # TODO(#38095): Replace assertEqualIgnoreType. See issue #38095
         self.assertEqualIgnoreType(complex_tensor.copy_imag(), imag)
 
-    def test_dtype_inference(self):
+    @dtypes(torch.float32, torch.float64)
+    def test_dtype_inference(self, device, dtype):
         # issue: https://github.com/pytorch/pytorch/issues/36834
-        torch.set_default_dtype(torch.double)
-        x = torch.tensor([3., 3. + 5.j])
-        self.assertEqual(x.dtype, torch.cdouble)
+        torch.set_default_dtype(dtype)
```

Comment:
If you set the default dtype you need to unset it (it's a global setting).

try:
  capture current default dtype
  set default dtype
  run test
finally:
  reset default dtype
```

Really we should write a default dtype context manager and put it in the common test utils to make this easier.

Reply:
Fixed